packages/foundry/contracts/variable/VRWitch.sol

```diff
@@ -0,0 +1,152 @@
+    /// @param vaultId Id of the vault to liquidate
+    function _auctionStarted(
+        bytes12 vaultId,
+        DataTypes.Auction memory auction_,
+        DataTypes.Line memory line
+    ) internal virtual returns (VRDataTypes.Vault memory vault) {
+        // The Witch is now in control of the vault under auction
+        vault = IVRCauldron(address(cauldron)).give(vaultId, address(this));
+        emit Auctioned(
+            vaultId,
+            auction_,
+            line.duration,
+            line.collateralProportion
+        );
+    }
+
+    // ======================================================================
+    // =                          Bidding functions                         =
+    // ======================================================================
+
+    /// @notice Returns debt that could be paid given the maxBaseIn
+    function _debtFromBase(DataTypes.Auction memory auction_, uint128 maxBaseIn)
+        internal
+        override
+        returns (uint256 artIn)
+    {
+        artIn = cauldron.debtFromBase(auction_.baseId, maxBaseIn);
+    }
+
+    /// @notice Returns base that could be paid given the artIn
+    function _debtToBase(DataTypes.Auction memory auction_, uint128 artIn)
+        internal
+        override
+        returns (uint256 baseIn)
+    {
+        baseIn = cauldron.debtToBase(auction_.baseId, artIn);
+    }
+
+    // ======================================================================
+    // =                         Quoting functions                          =
+    // ======================================================================
+
+    /// @dev quotes how much ink a liquidator is expected to get if it repays an `artIn` amount. Works for both Auctioned and ToBeAuctioned vaults
+    /// @param vaultId The vault to get a quote for
+    /// @param to Address that would get the collateral bought
+    /// @param maxArtIn How much of the vault debt will be paid. GT than available art means all
+    /// @return liquidatorCut How much collateral the liquidator is expected to get
+    /// @return auctioneerCut How much collateral the auctioneer is expected to get. 0 if liquidator == auctioneer
+    /// @return artIn How much debt the liquidator is expected to pay
+    function calcPayout(
+        bytes12 vaultId,
+        address to,
+        uint256 maxArtIn
+    )
+        external
+        view
+        override
+        returns (
+            uint256 liquidatorCut,
+            uint256 auctioneerCut,
+            uint256 artIn
+        )
+    {
+        DataTypes.Auction memory auction_ = auctions[vaultId];
+        VRDataTypes.Vault memory vault = IVRCauldron(address(cauldron)).vaults(
+            vaultId
+        );
+
+        // If the vault hasn't been auctioned yet, we calculate what values it'd have if it was started right now
+        if (auction_.start == 0) {
+            DataTypes.Balances memory balances = cauldron.balances(vaultId);
```

Comment:
if you had a template method that returns the values you get from the vault and/or series, then you can avoid duplicating this whole method. You may also be able to re-use that method to simplify more code and/or remove even more duplication on other places